cupy/random/generator.py

```diff
@@ -948,13 +948,17 @@ def shuffle(self, a):
         if a.ndim == 0:
             raise TypeError('An array whose ndim is 0 is not supported')
 
-        a[:] = a[self.permutation(len(a))]
+        a[:] = a[self._permutation(len(a))]
 
-    def permutation(self, num):
-        """Returns a permuted range."""
-        if not isinstance(num, six.integer_types):
-            raise TypeError('The data type of argument "num" must be integer')
+    def permutation(self, a):
+        """Returns a permuted range or a permutation of an array."""
+        if isinstance(a, six.integer_types):
+            return self._permutation(a)
+        else:
+            return a[self._permutation(len(a))]
```

Comment:
Should we allow permutations along a different axis?
This only does a permutation of the first dimension.

Reply:
> Should we allow permutations along a different axis?
> This only does a permutation of the first dimension.

It's compatible with NumPy.  https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.permutation.html

> If *x* is a multi-dimensional array, it is only shuffled along its first index.

Reply:
Perfect then!